Add basics 1 to implement 1

schettn opened the pull request:
**Checklist:**

 <!--- Make sure your PR is documented and tested before submission. Put an `x` in all the boxes that apply: -->
 - [x] Have you added an explanation of what your changes do and why you'd like them to be included?
 - [x] Have you updated or added documentation for the change?
 - [x] Have you tested your changes with successful results?


**Type of Changes:**

 <!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
 - [ ] Bug fix (non-breaking change which fixes an issue)
 - [ ] New feature (non-breaking change which adds functionality)
 - [x] Breaking change (fix or feature that would cause existing functionality to change)


**What is the current behavior? (link to any open issues here)**
- Currently the GraphQL client cannot handle mutations. #6 
- Currently the snek-client is not well documented. #7 
- Currently there is no way to fetch a website's DOM. #8 
- Currently there is no way to fetch a JSON file of a website. #9 
- Currently there are no templates for recurring tasks. #10 
- Currently there is no session awareness within the tasks and client in general. #11 
- Currently there are no tools for cookie management. #12


**What is the new behavior (if this is a feature change)?**
- Now the GraphQL client can handle queries and mutations. It supports easy integration
   of any further types. Closes #6
- Now the snek-client is documented. Closes #7
- Now it is possible to fetch a website DOM by its url. Closes #8
- Now it is possible to fetch a JSON file of a website by its url. Closes #9
- Now there are templates for recurring tasks. Closes #10
- Now the tasks are session aware. The client can handle sessions. Closes #11
- Now there are multiple tools for cookie management. Closes #12

schettn committed:
Add basic client structure

The basic client structure has been adopted from snek-eventmanager.

schettn committed:
Expand client structure

Important features and type declarations have been added.

schettn committed:
Code cleanup

Some comments are added now.

schettn committed:
Differ clients

Add GithubClient and SnekClient inherited from Client.

schettn committed:
Enhance session refresh

Now when the token is not set it will use the token from the cookies.

schettn committed:
Add username to auth process

Now the username is added to the auth mutation and UserData
interface.

schettn committed:
Relocate whoami task

The whoami task is located in the user task now.

schettn committed:
Tasks code cleanup

The tasks have refined return types now.

schettn committed:
Add allPageUrls task and query

A task with query which can fetch all page urls is implemented now.

schettn committed:
Fix registration query typo

The incorrect registration query is fixed now.

schettn committed:
Implement SnekGqlTemplate interface

A interface for SnekGqlTemplate is implemented now.

schettn committed:
Code cleanup

Some comments are added now.

schettn committed:
Redefine endpoints

Split the endpoint interface into multiple ones.

schettn committed:
Move send method in sub session

Now the send method is not longer in the main session but in
the snek sub session

schettn committed:
Implement scraper endpoint

Now the scraper endpoint is added.

schettn committed:
Add gitlab client

Now the gitlab client is added.

schettn committed:
Fix not defined errors

Due to the refining of the sessions and addSubSession method there were
some errors in the code. These are fixed now.

schettn committed:
Refine getDom in scraper

Now the getDom returns a Document object.

schettn committed:
Add send method to github session

The send method is added to the github sub session.

schettn committed:
Add scrapers header

Headers are added to the scraper functions now.

schettn committed:
Add documentation

Some code documentation has been added.

schettn committed:
Adjust CacheData

The platformData in the user object is replaced with an id now.

schettn committed:
Add is alive check

Adjust the isAlive check of the super to check
token and refresh token.

schettn committed:
Set refresh token cookie max age

The max age date of the refresh token cookie has
been reduced for testing reasons. This will be changed in later steps.

schettn committed:
Replace cookie expire with max age

The cookie expire is repalced with max age now.

schettn committed:
Code cleanup

The code has been cleaned up.

schettn committed:
Add production exception

Lib has been added to be ignored.

schettn committed:
Remove lib

Lib should not be pushed therefore it has been
removed now.

schettn committed:
Add username to whoami

The username has been added to the whoami
query.

schettn committed:
Refine cookies

The max-age is reset to expires and the token TTL times have been refined.

schettn committed:
Refine cookies

Some times and have been refined. Also a removeCookie before a createCookie is implemented now.

schettn committed:
Fix end session issue

A issue where the the refreshToken was revoked
but his cookie not deleted is fixed now.

schettn committed:
Update profile query

The profile query and their task have been updated.
Now firstName, lastName and email are available.

schettn committed:
Improve code quality

The code quality has been improved an some
documentation has been added.

schettn committed:
Refine tsconfig

The documentation of the file has been refined.
Also the target is specified now.

pinterid commented on .npmignore:
```diff
@@ -0,0 +1,8 @@
+CODE_OF_CONDUCT.md
+CONTRIBUTING.md
+PULL_REQUEST_TEMPLATE
+README.md
```
Add newline.

pinterid commented on package.json:
```diff
@@ -1,42 +1,50 @@
   "keywords": [],
-  "author": "schettn",
-  "license": "EUPL-1.2",
+  "author": "",
```
Missing author.

pinterid commented on package.json:
```diff
@@ -1,42 +1,50 @@
-    "not ie <= 11",
-    "not op_mini all"
-  ]
+  "babel": {
```
Is this necessary?

pinterid commented on src/endpoints/apollo.ts:
```diff
@@ -0,0 +1,120 @@
+import {
+  InMemoryCache,
+  IntrospectionFragmentMatcher,
+  NormalizedCacheObject,
```
Comma is not necessary.

pinterid commented on src/endpoints/apollo.ts:
```diff
@@ -0,0 +1,120 @@
+    try {
+      this.cache = new InMemoryCache({ fragmentMatcher });
+    } catch {
+      throw new Error("An error occurred when initializing cache!");
```
Change to : An error occurred while initializing the cache!

pinterid commented on src/endpoints/apollo.ts:
```diff
@@ -0,0 +1,120 @@
+        headers: options.headers,
+      });
+    } catch {
+      throw new Error("An error occurred when initializing the API link!");
```
Change to : An error occurred while initializing the API link!

pinterid commented on src/endpoints/apollo.ts:
```diff
@@ -0,0 +1,120 @@
+        link: this.link,
+      });
+    } catch {
+      throw new Error("An error occurred when initializing headers!");
```
Change to : An error occurred while initializing the headers!

pinterid commented on src/endpoints/apollo.ts:
```diff
@@ -0,0 +1,120 @@
+    type: string,
+    data: DocumentNode,
+    variables?: object,
+    headers?: object
```
Missing comma.

schettn replied:
In this case a comma is not needed because this is not an object declaration over multiple lines. This are method arguments.

pinterid commented on src/endpoints/index.ts:
```diff
@@ -0,0 +1,72 @@
+   * Desc: A endpoint description.
+   */
+  desc: string;
+}
```
Missing new line.

pinterid commented on src/endpoints/index.ts:
```diff
@@ -0,0 +1,72 @@
+    type: string,
+    data: DocumentNode,
+    variables?: object,
+    headers?: object
```
Missing comma.

kleberbaum replied:
In this case a comma is not needed because this is not an object declaration over multiple lines. This are method arguments.

pinterid commented on src/endpoints/index.ts:
```diff
@@ -0,0 +1,72 @@
+/** @interface ScraperEndpoint defines the structure of the scraper endpoint. */
+interface ScraperEndpoint extends IEndpoint {
+  /**
+   * GetJson: A method which get gets json data from a specific url.
```
Remove get.

pinterid commented on src/endpoints/index.ts:
```diff
@@ -0,0 +1,72 @@
+   */
+  getJson<T>(url: string): Promise<T>;
+  /**
+   * GetDom: A method which get gets DOM data from a specific url.
```
Remove get.

pinterid commented on src/endpoints/scraper.ts:
```diff
@@ -0,0 +1,77 @@
+//#region > Imports
+//> Interfaces
+// Contains the interface for the Apollo endpoint and the Apollo options
```
The second use of "the Apollo" could be removed.

pinterid commented on src/endpoints/scraper.ts:
```diff
@@ -0,0 +1,77 @@
+// Contains the interface for the Apollo endpoint and the Apollo options
+import { ScraperEndpoint, IOptions } from "./index";
+//#endregion
+
```
Missing: //#region > Classes

pinterid commented on src/endpoints/scraper.ts:
```diff
@@ -0,0 +1,77 @@
+      });
+  }
+}
+
```
Missing: #//endregion
ends the Class region

pinterid commented on src/index.ts:
```diff
@@ -0,0 +1,92 @@
+  type: string;
+  headers: object;
+}
+/** @interface Client will define the client structure. */
```
Missing new line

pinterid commented on src/index.ts:
```diff
@@ -0,0 +1,92 @@
+  constructor(
+    type: string = "graphql",
+    url: string = "https://engine.snek.at/api/graphiql",
+    headers: object = {}
```
Missing comma.

schettn replied:
In this case a comma is not needed because this is not an object declaration over multiple lines. This are constructor arguments.

pinterid commented on src/index.ts:
```diff
@@ -0,0 +1,92 @@
+  constructor(
+    url: string = "https://api.github.com/graphql",
+    type: string = "graphql",
+    headers: object = {}
```
Missing comma.

schettn replied:
In this case a comma is not needed because this is not an object declaration over multiple lines. This are constructor arguments.

pinterid commented on package.json:
```diff
@@ -1,42 +1,50 @@
-  "author": "schettn",
-  "license": "EUPL-1.2",
+  "author": "",
+  "license": "ISC",
```
Wrong license?

pinterid commented on src/index.ts:
```diff
@@ -0,0 +1,92 @@
+  constructor(
+    url: string = "https://gitlab.com",
+    type: string = "scraper",
+    headers: object = {}
```
Missing comma.

schettn replied:
In this case a comma is not needed because this is not an object declaration over multiple lines. This are constructor arguments.

pinterid commented on src/session/cookie-utils.ts:
```diff
@@ -0,0 +1,72 @@
+  const date = new Date();
+  const value = val;
+
+  // Set it expire in n(time) seconds
```
Change to "Set its expire in n(time) seconds"

Aichnerc replied:
I'd rather write "Set cookie expiry date in n(time) seconds".

pinterid commented on src/session/cookie-utils.ts:
```diff
@@ -0,0 +1,72 @@
+function deleteCookie(name: string) {
+  const date = new Date();
+
+  // Set it expire in -1 days
```
Change to "Set its expire in -1 days"

Aichnerc replied:
I'd rather write "Set cookie expiry date to -1 days".

pinterid commented on src/session/cookie-utils.ts:
```diff
@@ -0,0 +1,72 @@
+/** @description General utils for managing cookies in Typescript. */
```
Add //#region > Functions

pinterid commented on src/session/cookie-utils.ts:
```diff
@@ -0,0 +1,72 @@
+  }
+  return false;
+}
+
```
//#endregion

pinterid commented on src/session/index.ts:
```diff
@@ -0,0 +1,90 @@
+//#endregion
+
+//#region > Interfaces
+/** @interface UserData defines the structure a profile objects. */
```
Change to : "UserData defines the structure of profile objects."

pinterid commented on src/session/index.ts:
```diff
@@ -0,0 +1,90 @@
+  lastLogin?: string;
+}
+
+/** @interface UserData defines the structure a authentication object. */
```
Change to: "UserData defines the structure of an authentication object."

pinterid commented on src/session/index.ts:
```diff
@@ -0,0 +1,90 @@
+  refreshToken: string;
+}
+
+/** @interface User defines the structure a basic user object. */
```
Change to: "User defines the structure of a basic user object."

pinterid commented on src/session/index.ts:
```diff
@@ -0,0 +1,90 @@
+
+  //> Methods
+  /**
+   * Add a  subSession.
```
Remove a space after "a".

pinterid commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+// Contains basic session interfaces
+import { IAuth, User, UserData } from "./index";
+//#endregion
+
```
Missing //#region > Classes

pinterid commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+  constructor(
+    sId: string,
+    public ep: ApolloEndpoint,
+    public template: IMainTemplate
```
Missing comma.

schettn replied:
In this case a comma is not needed because this is not an object declaration over multiple lines. This are constructor arguments.

pinterid commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+  }
+
+  /**
+   * Sned query:
```
Change "Sned" to "Send".

pinterid commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+   * @description Send a query to the endpoint.
+   * @param {string} token A authentication token.
+   * @param {DocumentNode} data A DocumentNode with a query.
+   * @param {object} variables A abject with variables.
```
Change to "variables A object with variables."

pinterid commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+   * Is alive check.
+   *
+   * @description Token and refresh token status check.
+   * @param {boolean} alive A status whether the token and refresh token is alive or not
```
Change to "are alive"

pinterid commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+    }
+  }
+
+  /** @description End session by reset jwt and deleting cookie. */
```
Change to "End session by resetting jwt and deleting cookies."

pinterid commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+    deleteCookie(this.refreshTokenName);
+  }
+}
+
```
//#endregion 
end the region for classes

pinterid commented on src/templates/index.ts:
```diff
@@ -0,0 +1,28 @@
+//#endregion
+
+//#region > Classes
+/** @class A Template with initializes all SubTemplates. */
```
Change to "A Template which initializes all SubTemplates."

pinterid commented on src/templates/snek/gql/index.ts:
```diff
@@ -0,0 +1,32 @@
+//#endregion
+
+//#region > Classes
+/** @class A Template with initializes all SubTemplates. */
```
Change to: "A Template which initializes all SubTemplates."

pinterid commented on src/templates/snek/gql/mutations/index.ts:
```diff
@@ -0,0 +1,23 @@
+import * as _jwtAuth from "./jwtAuth";
+// Contains the user mutations
+import * as _user from "./user";
+
```
Missing //#endregion

pinterid commented on src/templates/snek/gql/mutations/index.ts:
```diff
@@ -0,0 +1,23 @@
+import * as _user from "./user";
+
+//#region > Classes
+/** @class A Template with initializes all mutations. */
```
Change to: "A Template which initializes all mutations."

pinterid commented on src/templates/snek/gql/mutations/jwtAuth.ts:
```diff
@@ -0,0 +1,86 @@
+    }
+  }
+`;
+/**
```
Missing new line.

pinterid commented on src/templates/snek/gql/mutations/user.ts:
```diff
@@ -0,0 +1,57 @@
+    }
+  }
+`;
+/**
```
Missing new line.

pinterid commented on src/templates/snek/gql/queries/index.ts:
```diff
@@ -0,0 +1,21 @@
+//#endregion
+
+//#region > Classes
+/** @class A template with initializes all SubTemplates. */
```
Change to "A template which initializes all SubTemplates."

pinterid commented on src/templates/snek/gql/queries/index.ts:
```diff
@@ -0,0 +1,21 @@
+
+//#region > Exports
+export { SnekGqlQuery };
+
```
Missing //#endregion

pinterid commented on src/templates/snek/gql/queries/user.ts:
```diff
@@ -0,0 +1,62 @@
+      lastLogin
+    }
+  }
+`;
```
Missing new line.

pinterid commented on src/templates/snek/gql/tasks/auth.ts:
```diff
@@ -0,0 +1,153 @@
+        username: "cisco",
+        password: "ciscocisco",
+      }
+    );
```
Missing new line.

pinterid commented on src/templates/snek/gql/tasks/general.ts:
```diff
@@ -0,0 +1,90 @@
+
+//#region > Exports
+export default SnekGqlGeneralTasks;
+
```
Missing //#endregion

pinterid commented on src/templates/snek/gql/tasks/index.ts:
```diff
@@ -0,0 +1,43 @@
+interface IResponse {
+  errors: [];
+}
+
```
Missing //#endregion

pinterid commented on src/templates/snek/gql/tasks/user.ts:
```diff
@@ -0,0 +1,179 @@
+interface WhoamiData {
+  username: string;
+}
+
```
Missing //#endregion

pinterid commented on src/templates/snek/index.ts:
```diff
@@ -0,0 +1,19 @@
+//#endregion
+
+//#region > Classes
+/** @class A Template with initializes all SubTemplates. */
```
Chane to: "A Template which initializes all SubTemplates."

pinterid reviewed (changes requested):
Modify before merge!

Aichnerc commented on src/session/cookie-utils.ts:
```diff
@@ -0,0 +1,72 @@
+
+  // Set it expire in n(time) seconds
+  date.setTime(date.getTime() + time * 1000);
+  // Set it
```
Change to "Create cookie".

Aichnerc commented on src/session/cookie-utils.ts:
```diff
@@ -0,0 +1,72 @@
+  // Set it expire in -1 days
+  date.setTime(date.getTime() + -1 * 24 * 60 * 60 * 1000);
+
+  // Set it
```
Change to "Create cookie".

Aichnerc commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+    this.token = auth.token;
+    this.refreshToken = auth.refreshToken;
+
+    /** Delte the token and refreshToken cookie. */
```
Typo: Delte -> Delete

Aichnerc commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+    let response;
+
+    if (!user && this.wasAlive()) {
+      /** Refresh tokens. */
```
Please consider using our comment style for single line comments.
`// Comment`
Single line comments shall not have a period if it contains less than two sentences.

Aichnerc commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+      this.refresh();
+    } else {
+      if (!user) {
+        /** Authenticate anonymous user. */
```
Please consider using our comment style for single line comments.
`// Comment`
Single line comments shall not have a period if it contains less than two sentences.

Aichnerc commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+        /** Authenticate anonymous user. */
+        response = await this.tasks.auth.anon();
+      } else {
+        /** Authenticate real user. */
```
Please consider using our comment style for single line comments.
`// Comment`
Single line comments shall not have a period if it contains less than two sentences.

Aichnerc commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+        throw new Error(JSON.stringify(response.errors));
+      }
+
+      /** Set tokens. */
```
Please consider using our comment style for single line comments.
`// Comment`
Single line comments shall not have a period if it contains less than two sentences.

Aichnerc commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+      return <UserData>response.data.auth.user;
+    }
+
+    /** Get user data. */
```
Please consider using our comment style for single line comments.
`// Comment`
Single line comments shall not have a period if it contains less than two sentences.

Aichnerc commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+  async refresh() {
+    if (!this.isAlive()) {
+      if (this.wasAlive()) {
+        /**
+         * Refresh token with refreshToken
+         */
```
Please consider using our comment style for single line comments.
`// Comment`

Aichnerc commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+
+        this.initTokens(response.data.refresh);
+      } else {
+        /** Begin new session */
```
Please consider using our comment style for single line comments.
`// Comment`

Aichnerc commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+       */
+      console.log("TID-1(REVOKE)", response.data.revoke.revoked);
+    }
+    /** Reset token. */
```
Please consider using our comment style for single line comments.
`// Comment`

Aichnerc commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+
+  /** @description End session by reset jwt and deleting cookie. */
+  async end() {
+    /** Revoke token if it is set */
```
Please consider using our comment style for single line comments.
`// Comment`

Aichnerc commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,249 @@
+    this.token = "";
+    this.refreshToken = "";
+
+    /** Delete cookie. */
```
Please consider using our comment style for single line comments.
`// Comment`

Aichnerc reviewed (changes requested):
Should be okay after a few minor changes regaring commenting style.

schettn committed:
Remove empty file

A unnecessary file is removed now.

schettn committed:
Refine package.json

The description, author, licence, devDependencies and bable
settings have been refined due to a request of @pinterid.

simonprast committed:
Improve comments

Change some comment contents and comment styles

pinterid committed:
Improve code quality

The code quality has been improved an some
documentation has been added.

pinterid reviewed (approved):
Should be good to merge.

schettn committed:
Improve code quality

The comments of the tsconfig have been refined.

schettn committed:
Remove include setting

The include setting is removed because of this would exclude
needed files. Therefore we choose the default settings.

Aichnerc reviewed (approved):
Looks good to me. 👀

simonprast reviewed (approved):
Looks twice the nice

kleberbaum commented on package.json:
```diff
@@ -1,12 +1,16 @@
   "version": "1.0.0",
-  "main": "index.js",
-  "type": "module",
+  "description": "Enoy it.",
```
xD

schettn replied:
I would leave it that way. Represents my feelings.

kleberbaum commented on src/endpoints/apollo.ts:
```diff
@@ -0,0 +1,120 @@
+  private client: ApolloClient<NormalizedCacheObject>;
+
+  headers: object;
+  desc: string = "Apollo Endpoint";
```
client

kleberbaum commented on src/endpoints/index.ts:
```diff
@@ -0,0 +1,72 @@
+/** @interface ApolloEndpoint defines the structure of the apollo endpoint. */
+interface ApolloEndpoint extends IEndpoint {
+  /**
+   * Send: A method which provides send functionallity for all graphql types.
```
reword

kleberbaum commented on src/endpoints/scraper.ts:
```diff
@@ -0,0 +1,77 @@
+import { ScraperEndpoint, IOptions } from "./index";
+//#endregion
+
+/**@class A endpoint to fetch website page data. */
```
Add a space after ```/** ```.

kleberbaum replied:
```A endpoint to fetch page DOM.```

kleberbaum commented on src/index.ts:
```diff
@@ -0,0 +1,92 @@
+//#endregion
+
+//#region > Classes
+/**@class The snek-client. Enjoy it. Will be implemented in the future. */
```
Add a space after ```/** ```.

kleberbaum commented on src/index.ts:
```diff
@@ -0,0 +1,92 @@
+  constructor(ep: IEndpoint) {}
+}
+
+/**@class A client implementation for snek interaction */
```
Add a space after ```/** ```.

kleberbaum replied:
Please add a period.

kleberbaum commented on src/endpoints/scraper.ts:
```diff
@@ -0,0 +1,77 @@
+class Scraper implements ScraperEndpoint {
+  //> Fields
+  headers: object;
+  desc: string = "A endpoint to fetch website page data.";
```
```A endpoint to fetch page DOM.```

kleberbaum commented on src/endpoints/scraper.ts:
```diff
@@ -0,0 +1,77 @@
+  /**
+   * @constructor
+   * @author Nico Schett <contact@schett.net>
+   * @param root Root url of endpoint. Specify it like https://foo.bar. Correct slash setting is IMPORTANT!
```
IMPORTANT! This constructor shall only be used  in the right set and setting.

kleberbaum commented on src/endpoints/scraper.ts:
```diff
@@ -0,0 +1,77 @@
+
+  //> Methods
+  /**
+   * @param path Path to the endpoint. Specify it like "/foo/bar". Correct slash setting is IMPORTANT!
```
IMPORTANT! This constructor shall only be used in the right set and setting.

kleberbaum commented on src/endpoints/scraper.ts:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  /**
+   * @param path Path to the endpoint. Specify it like "/foo/bar". Correct slash setting is IMPORTANT!
```
IMPORTANT! This constructor shall only be used in the right set and setting.

kleberbaum commented on src/endpoints/scraper.ts:
```diff
@@ -0,0 +1,77 @@
+        ...this.headers,
+      },
+    })
+      .then((response) => {
```
Please check indentation.

schettn replied:
Checked and it is correct.
Ref: https://eslint.org/docs/rules/indent#memberexpression

kleberbaum commented on src/session/cookie-utils.ts:
```diff
@@ -0,0 +1,74 @@
+  const value = val;
+
+  // Set cookie expiry date in n(time) seconds
+  date.setTime(date.getTime() + time * 1000);
```
Please add a newline.

kleberbaum commented on src/session/cookie-utils.ts:
```diff
@@ -0,0 +1,74 @@
+
+  if (cookie) {
+    return true;
+  }
```
Please add a newline.

kleberbaum commented on src/session/index.ts:
```diff
@@ -0,0 +1,90 @@
+   * @description Add a subSession to a session.
+   */
+  addSubSession<S, E, T>(childSId: string, Cls: any, endpoint: E, template: T) {
+    let session: S = new Cls(this.sId + "_" + childSId, endpoint, template);
```
Please change to ```cls()``` lowercase, I am begging you.

Aichnerc replied:
He is refering to a class. Once it is a class name it should start with an uppercase letter.

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+//#endregion
+
+//#region > Classes
+/** @class A Github SubSession  */
```
```/** @class A Github SubSession */```

kleberbaum replied:
Please add a period.

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+  }
+}
+
+/** @class A Snek SubSession  */
```
```/** @class A Snek SubSession */```

kleberbaum replied:
Please add a period.

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+    public ep: ApolloEndpoint,
+    public template: IMainTemplate
+  ) {
+    /* TODO: Change template set to github */
```
?

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+
+    /* Delete the token and refreshToken cookie. */
+    deleteCookie(this.tokenName);
+    deleteCookie(this.refreshTokenName);
```
Please add a newline.

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+    let response;
+
+    if (!user && this.wasAlive()) {
+      /* Refresh tokens. */
```
Why is there a period?

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+    this.token = auth.token;
+    this.refreshToken = auth.refreshToken;
+
+    /* Delete the token and refreshToken cookie. */
```
Why is there a period?

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+    /* Delete the token and refreshToken cookie. */
+    deleteCookie(this.tokenName);
+    deleteCookie(this.refreshTokenName);
+    /* Set the token and refreshToken cookie with expire times. */
```
Why is there a period?

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+      this.refresh();
+    } else {
+      if (!user) {
+        /* Authenticate anonymous user. */
```
Why is there a period?

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+        /* Authenticate anonymous user. */
+        response = await this.tasks.auth.anon();
+      } else {
+        /* Authenticate real user. */
```
Why is there a period?

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+        throw new Error(JSON.stringify(response.errors));
+      }
+
+      /* Set tokens. */
```
Why is there a period?

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+      return <UserData>response.data.auth.user;
+    }
+
+    /* Get user data. */
```
Why is there a period?

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+      /* TID: 1 */
+      console.log("TID-1(REVOKE)", response.data.revoke.revoked);
+    }
+    /* Reset token. */
```
Why is there a period?

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+    this.token = "";
+    this.refreshToken = "";
+
+    /* Delete cookie. */
```
Why is there a period?

kleberbaum commented on tsconfig.json:
```diff
@@ -0,0 +1,24 @@
+{
+  "compilerOptions": {
+    "target": "ES2020",
+    /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'. */
```
Why is there a period?

kleberbaum commented on tsconfig.json:
```diff
@@ -0,0 +1,24 @@
+    "target": "ES2020",
+    /* Specify module code generation: 'none', 'commonjs', 'amd', 'system', 'umd', 'es2015', 'es2020', or 'ESNext'. */
+    "module": "commonjs",
+    /* Generates corresponding '.d.ts' file. */
```
Why is there a period?

kleberbaum commented on tsconfig.json:
```diff
@@ -0,0 +1,24 @@
+    "module": "commonjs",
+    /* Generates corresponding '.d.ts' file. */
+    "declaration": true,
+    /* Redirect output structure to the directory. */
```
Why is there a period?

kleberbaum commented on tsconfig.json:
```diff
@@ -0,0 +1,24 @@
+    "declaration": true,
+    /* Redirect output structure to the directory. */
+    "outDir": "lib",
+    /* Do not emit comments to output. */
```
Why is there a period?

kleberbaum commented on tsconfig.json:
```diff
@@ -0,0 +1,24 @@
+    "outDir": "lib",
+    /* Do not emit comments to output. */
+    "removeComments": true,
+    /* Enable all strict type-checking options. */
```
Why is there a period?

kleberbaum commented on tsconfig.json:
```diff
@@ -0,0 +1,24 @@
+      via creation of namespace objects for all imports. Implies 'allowSyntheticDefaultImports'.
+     */
+    "esModuleInterop": true,
+    /* Disallow inconsistently-cased references to the same file. */
```
Why is there a period?

kleberbaum commented on tsconfig.json:
```diff
@@ -0,0 +1,24 @@
+    /* Disallow inconsistently-cased references to the same file. */
+    "forceConsistentCasingInFileNames": true
+  },
+  /* The files to not type check. */
```
Why is there a period?

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+    if (!this.isAlive()) {
+      if (this.wasAlive()) {
+        /* Refresh token with refreshToken */
+        this.refreshToken = getCookie(this.refreshTokenName);
```
Please add a newline.

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+  async end() {
+    /* Revoke token if it is set */
+    if (this.refreshToken !== "") {
+      let response = await this.tasks.auth.revoke();
```
Please add a newline.

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+    if (this.refreshToken !== "") {
+      let response = await this.tasks.auth.revoke();
+      /* TID: 1 */
+      console.log("TID-1(REVOKE)", response.data.revoke.revoked);
```
Please make it into a comment and change title to //#DEBUG TSID1.

kleberbaum commented on src/templates/snek/gql/mutations/jwtAuth.ts:
```diff
@@ -0,0 +1,88 @@
+//#region > Imports
+//> Graphql
+// Contains a a gql tag for wrapping queries
+import gql from "graphql-tag";
```
```//#INSTALL "apollo-client"```

kleberbaum commented on src/endpoints/apollo.ts:
```diff
@@ -0,0 +1,120 @@
+//#region > Imports
+//> Apollo Client
+// Contains the client for graphql handeling
+import { ApolloClient } from "apollo-client";
```
```//#INSTALL "apollo-client"```

kleberbaum commented on src/endpoints/apollo.ts:
```diff
@@ -0,0 +1,120 @@
+import { ApolloClient } from "apollo-client";
+//> Apollo Link
+// Contains the link for the apollo client
+import { HttpLink } from "apollo-link-http";
```
```//#INSTALL "apollo-client"```

kleberbaum commented on src/endpoints/apollo.ts:
```diff
@@ -0,0 +1,120 @@
+  InMemoryCache,
+  IntrospectionFragmentMatcher,
+  NormalizedCacheObject,
+} from "apollo-cache-inmemory";
```
```//#INSTALL "apollo-client"```

kleberbaum commented on src/endpoints/apollo.ts:
```diff
@@ -0,0 +1,120 @@
+} from "apollo-cache-inmemory";
+//> Interfaces
+// Contains the interface for gql queries, mutations and subscriptions
+import { DocumentNode } from "graphql";
```
```//#INSTALL "apollo-client"```

kleberbaum commented on src/endpoints/index.ts:
```diff
@@ -0,0 +1,72 @@
+//#region > Imports
+//> Interfaces
+// Contains the interface for gql queries, mutations and subscriptions
+import { DocumentNode } from "graphql";
```
```//#INSTALL "apollo-client"```

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+import SnekTasks from "../templates/snek/gql/tasks/index";
+//> Interfaces
+// Contains the interface for gql queries, mutations and subscriptions
+import { DocumentNode } from "graphql";
```
```//#INSTALL "apollo-client"```

kleberbaum commented on src/templates/snek/gql/mutations/user.ts:
```diff
@@ -0,0 +1,58 @@
+//#region > Imports
+//> Graphql
+// Contains a a gql tag for wrapping queries
+import gql from "graphql-tag";
```
```//#INSTALL "apollo-client"```

kleberbaum commented on src/templates/snek/gql/queries/general.ts:
```diff
@@ -0,0 +1,54 @@
+//#region > Imports
+//> Graphql
+// Contains a a gql tag for wrapping queries
+import gql from "graphql-tag";
```
```//#INSTALL "apollo-client"```

kleberbaum commented on src/templates/snek/gql/queries/user.ts:
```diff
@@ -0,0 +1,63 @@
+//#region > Imports
+//> Graphql
+// Contains a a gql tag for wrapping queries
+import gql from "graphql-tag";
```
```//#INSTALL "apollo-client"```

kleberbaum commented on src/templates/snek/gql/mutations/jwtAuth.ts:
```diff
@@ -0,0 +1,88 @@
+
+/**
+ * JWT Revoke:
+ * Mutation to verify a JWT
```
```* Mutation to verify a JWT```

schettn replied:
I am begging you to present a suggestion for improvement correctly. It doesn't help me if you point your finger at a spot but don't say what bothers you. Thanks.

kleberbaum commented on src/templates/snek/gql/queries/user.ts:
```diff
@@ -0,0 +1,63 @@
+ * Query to fetch profile data
+ * 
+ * @param {string} url PageUrl of a user profile
+ * @param {string} token  A users JWT
```
One whitespace too much.

kleberbaum commented on src/templates/snek/gql/queries/user.ts:
```diff
@@ -0,0 +1,63 @@
+ * @param {string} url PageUrl of a user profile
+ * @param {string} token  A users JWT
+ * @returns {string} Profile page of a user
+*/
```
Please fix indentation.

kleberbaum commented on src/templates/snek/gql/tasks/auth.ts:
```diff
@@ -0,0 +1,154 @@
+  refreshToken: string;
+}
+
+/** @interface RevokeResponse defines the structure of the refresh data. */
```
?

kleberbaum commented on src/templates/snek/gql/tasks/auth.ts:
```diff
@@ -0,0 +1,154 @@
+  data: { revoke: RevokeData };
+}
+
+/** @interface RevokeData defines the structure of the revoke data. */
```
?

kleberbaum commented on src/session/cookie-utils.ts:
```diff
@@ -0,0 +1,74 @@
+
+/**
+ * @function
+ * @returns {boolean} Is the cookie alive?.
```
?

kleberbaum commented on src/session/index.ts:
```diff
@@ -0,0 +1,90 @@
+   * Add a subSession.
+   *
+   * @param childSId The session name of the child.
+   * @param {any} type Specify the session (Session | string)
```
Please add a period.

kleberbaum commented on src/session/sessions.ts:
```diff
@@ -0,0 +1,243 @@
+   * @constructor
+   * @param {string} sId A session name
+   * @param {Endpoint} ep A endpoint
+   * @param {IMainTemplate} template A template set
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/mutations/jwtAuth.ts:
```diff
@@ -0,0 +1,88 @@
+ *
+ * @param {string} username Username of a user
+ * @param {string} password Password of a user
+ * @returns {string} JWT
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/mutations/jwtAuth.ts:
```diff
@@ -0,0 +1,88 @@
+ * Mutation to refresh a JWT
+ *
+ * @param {string} token A users JWT
+ * @returns {string} Refresh conformation
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/mutations/jwtAuth.ts:
```diff
@@ -0,0 +1,88 @@
+ * Mutation to verify a JWT
+ *
+ * @param {string} token A users JWT
+ * @returns {string} Verify conformation
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/mutations/jwtAuth.ts:
```diff
@@ -0,0 +1,88 @@
+ * Mutation to verify a JWT
+ *
+ * @param {string} token A users JWT
+ * @returns {string} Revoke conformation
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/mutations/user.ts:
```diff
@@ -0,0 +1,58 @@
+ *
+ * @param {string} token A users JWT
+ * @param {string} values Registration parameters defined by registration page
+ * @returns {string} Registration conformation
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/mutations/user.ts:
```diff
@@ -0,0 +1,58 @@
+ *
+ * @param {string} token A users JWT
+ * @param {string} platformData A serialized JSON object of all generated user data
+ * @returns {string} PlatformData of a user
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/queries/general.ts:
```diff
@@ -0,0 +1,54 @@
+ * Query to fetch all available GitLab server
+ *
+ * @param {string} JWT A users JWT
+ * @returns {string} A serialized JSON object with a list of GitLab server
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/queries/general.ts:
```diff
@@ -0,0 +1,54 @@
+ * Query to fetch all pages urls
+ *
+ * @param {string} JWT A users JWT
+ * @returns {string} A serialized JSON object with a list of all page urls
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/queries/user.ts:
```diff
@@ -0,0 +1,63 @@
+ * 
+ * @deprecated since 1.0 use jwtAuth instead
+ * @param {string} token A users JWT
+ * @returns {string} Username
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/queries/user.ts:
```diff
@@ -0,0 +1,63 @@
+ * 
+ * @param {string} url PageUrl of a user profile
+ * @param {string} token  A users JWT
+ * @returns {string} Profile page of a user
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/tasks/auth.ts:
```diff
@@ -0,0 +1,154 @@
+   *
+   * @constructor
+   * @author Nico Schett <contact@schett.net>
+   * @param {string} session A session for the tasks
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/tasks/auth.ts:
```diff
@@ -0,0 +1,154 @@
+  /**
+   * User login.
+   *
+   * @param {string} user A User defined by username and password
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/tasks/auth.ts:
```diff
@@ -0,0 +1,154 @@
+  /**
+   * Refresh token.
+   *
+   * @param {string} user A User defined by username and password
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/tasks/auth.ts:
```diff
@@ -0,0 +1,154 @@
+  /**
+   * Refresh token.
+   *
+   * @param {string} user A User defined by username and password
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/tasks/general.ts:
```diff
@@ -0,0 +1,91 @@
+//#endregion
+
+//#region > Classes
+/** @class A set of session aware Tasks */
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/tasks/general.ts:
```diff
@@ -0,0 +1,91 @@
+   *
+   * @constructor
+   * @author Nico Schett <contact@schett.net>
+   * @param {string} session A session for the tasks
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/tasks/general.ts:
```diff
@@ -0,0 +1,91 @@
+   * @returns {Promise<IGitlabServerResponse>} A list of Gitlab server.
+   */
+  async gitlabServer(): Promise<IGitlabServerResponse> {
+    /**
```
```/*```

kleberbaum commented on src/templates/snek/gql/tasks/general.ts:
```diff
@@ -0,0 +1,91 @@
+   * @returns {Promise<IAllPageUrlResponse>} A list of all page urls.
+   */
+  async allPageUrls(): Promise<IAllPageUrlResponse> {
+    /**
```
```/*```

kleberbaum commented on src/templates/snek/gql/tasks/user.ts:
```diff
@@ -0,0 +1,180 @@
+   *
+   * @constructor
+   * @author Nico Schett <contact@schett.net>
+   * @param {string} session A session for the tasks
```
Please add a period.

kleberbaum commented on src/templates/snek/gql/tasks/user.ts:
```diff
@@ -0,0 +1,180 @@
+   * @returns {Promise<IRegistrationResponse>} A JWT token.
+   */
+  async registration(values: object): Promise<IRegistrationResponse> {
+    /**
```
```/*```

kleberbaum commented on src/templates/snek/gql/tasks/user.ts:
```diff
@@ -0,0 +1,180 @@
+   * @returns {Promise<ICacheResponse>} A JWT token.
+   */
+  async cache(platformData: string): Promise<ICacheResponse> {
+    /**
```
```/*```

kleberbaum commented on src/templates/snek/gql/tasks/user.ts:
```diff
@@ -0,0 +1,180 @@
+   * @returns {Promise<IProfileResponse>} The page profile of a user.
+   */
+  async profile(url: string): Promise<IProfileResponse> {
+    /**
```
```/*```

kleberbaum commented on src/templates/snek/gql/tasks/user.ts:
```diff
@@ -0,0 +1,180 @@
+   * @returns {Promise<IWhoamiResponse>} User data.
+   */
+  async whoami(): Promise<IWhoamiResponse> {
+    /**
```
```/*```

kleberbaum commented on src/templates/snek/gql/tasks/auth.ts:
```diff
@@ -0,0 +1,154 @@
+   * Refresh token.
+   *
+   * @param {string} user A User defined by username and password
+   * @returns {Promise<IRefreshResponse>} A JWT token,
```
comma?

kleberbaum commented on src/templates/snek/gql/tasks/auth.ts:
```diff
@@ -0,0 +1,154 @@
+  }
+
+  /**
+   * Refresh token.
```
Revoke

kleberbaum commented on src/templates/snek/gql/tasks/auth.ts:
```diff
@@ -0,0 +1,154 @@
+  /**
+   * Refresh token.
+   *
+   * @param {string} user A User defined by username and password
```
No user needed.

schettn replied:
Wrong. `user` specifies the variable name. 

`A User defined by username and password` specifies the comment.
So this is completely intentional

kleberbaum commented on src/templates/snek/gql/tasks/auth.ts:
```diff
@@ -0,0 +1,154 @@
+  /**
+   * Refresh token.
+   *
+   * @param {string} user A User defined by username and password
```
No User needed.

schettn replied:
Wrong. `user` specifies the variable name. 

`A User defined by username and password` specifies the comment.
So this is completely intentional

kleberbaum commented on src/templates/snek/gql/tasks/general.ts:
```diff
@@ -0,0 +1,91 @@
+//#region > Classes
+/** @class A set of session aware Tasks */
+class SnekGqlGeneralTasks {
+  public template: ISnekGqlTemplate;
```
Please add a newline.

kleberbaum commented on src/templates/snek/gql/tasks/user.ts:
```diff
@@ -0,0 +1,180 @@
+  data: CacheData;
+}
+
+/** @interface CacheResponse defines the structure of a data response. */
```
CacheData

kleberbaum commented on src/templates/snek/gql/tasks/user.ts:
```diff
@@ -0,0 +1,180 @@
+//#region > Classes
+/** @class A set of session aware Tasks */
+class SnekGqlUserTasks {
+  public template: ISnekGqlTemplate;
```
Please add a newline.

kleberbaum commented on src/templates/snek/gql/tasks/user.ts:
```diff
@@ -0,0 +1,180 @@
+    await this.session.refresh();
+
+    let query = this.template.mutations.user.registration;
+    console.log(query);
```
Please remove from code.

kleberbaum commented on src/templates/snek/gql/tasks/user.ts:
```diff
@@ -0,0 +1,180 @@
+        token: this.session.token,
+        values,
+      }
+    );
```
Please add a newline.

kleberbaum commented on src/templates/snek/gql/tasks/user.ts:
```diff
@@ -0,0 +1,180 @@
+        token: this.session.token,
+        platformData,
+      }
+    );
```
Please add a newline.

kleberbaum reviewed (changes requested):
Should be ok to merge after a few changes.

kleberbaum committed:
Improve code quality

The code quality has been improved due to a
request of @pinterid.

schettn committed:
Improve code quality

The code quality has been improved due to a request of @kleberbaum.

schettn committed:
Adjust line length

A max. line limit was exceeded. That is fixed now.

kleberbaum reviewed (changes requested):
Should be ok to merge after a few changes.

schettn committed:
Improve code quality

The code quality has been improved due to a request of @kleberbaum.

kleberbaum commented on src/templates/snek/gql/mutations/jwtAuth.ts:
```diff
@@ -0,0 +1,89 @@
+ *
+ * @param {string} username Username of a user.
+ * @param {string} password Password of a user.
+ * @returns {string} JWT
```
```@returns {string} A JWT token.```

kleberbaum reviewed (approved):
Should be ok to merge.

schettn committed:
Improve code quality

The code quality has been improved due to the codacy check.

schettn committed:
Resolved merge conflict by giving full preference to add-basics-1.
